### PR TITLE
Add join_strings() call and use it from mattrs()

### PR DESCRIPTION
**Pull request by steven-johnson**
This is a super-nit kind of fix, but the fact that we had rerolled a join-strings algo in a half-dozen places made my teeth hurt, so I decided to fix it:

- Add join_strings() to Util.h
- revise the mattrs() calls to use it instead of the janky mess they used

This doesn't move the needle on code size or speed but it is less weird.

Probably other places we could/should use this too.

(Does C++20 have join/split strings in the std library yet? If not, why not?)

### CODE REVIEW

**Comment by abadams** [comment]
I've been wondering if we should consider stringstream an anti-pattern. If we know we're joining strings, this is about 3x faster:

```
std::string join_strings(const std::vector<std::string> &sources, const std::string &delim) {
    size_t sz = delim.size() * (sources.size() - 1);
    for (const auto &s : sources) {
        sz += s.size();
    }
    std::string result;
    result.reserve(sz);
    bool need_delim = false;
    for (const auto &s : sources) {
        if (need_delim) {
            result += delim;
        }
        result += s;
        need_delim = true;
    }
    return result;
}
```

**Comment by steven-johnson** [comment]
> I've been wondering if we should consider stringstream an anti-pattern. If we know we're joining strings, this is about 3x faster:
> 
> ```
> std::string join_strings(const std::vector<std::string> &sources, const std::string &delim) {
>     size_t sz = delim.size() * (sources.size() - 1);
>     for (const auto &s : sources) {
>         sz += s.size();
>     }
>     std::string result;
>     result.reserve(sz);
>     bool need_delim = false;
>     for (const auto &s : sources) {
>         if (need_delim) {
>             result += delim;
>         }
>         result += s;
>         need_delim = true;
>     }
>     return result;
> }
> ```

Abseil claims their StrCat() function is much faster for this reason. I'd be amenable to upgrading all this if we have data.

**Comment by abadams** [comment]
The 3x faster is a measured runtime on my workstation.

**Comment by steven-johnson** [comment]
> size_t sz = delim.size() * (sources.size() - 1);
>     for (const auto &s : sources) {
>         sz += s.size();
>     }
>     std::string result;
>     result.reserve(sz);
>     bool need_delim = false;
>     for (const auto &s : sources) {
>         if (need_delim) {
>             result += delim;
>         }
>         result += s;
>         need_delim = true;
>     }
>     return result;

well ok then, pushed (with a fix for zero-len array)

File: src/Util.h [hunk 1]
```diff
@@ -185,6 +186,29 @@ std::string replace_all(const std::string &str, const std::string &find, const s
 /** Split the source string using 'delim' as the divider. */
 std::vector<std::string> split_string(const std::string &source, const std::string &delim);
 
+/** Join the source vector using 'delim' as the divider. */
+template<typename T>
+std::string join_strings(const std::vector<T> &sources, const std::string &delim) {
+    size_t sz = 0;
+    if (!sources.empty()) {
+        delim.size() * (sources.size() - 1);
```
Comment (abadams) [inline src/Util.h:194]: Should this be sz = ?

Comment (steven-johnson) [reply src/Util.h:194]: oops

**Review by abadams — APPROVED** [review]

File: src/Util.h [hunk 2]
```diff
@@ -185,6 +186,29 @@ std::string replace_all(const std::string &str, const std::string &find, const s
 /** Split the source string using 'delim' as the divider. */
 std::vector<std::string> split_string(const std::string &source, const std::string &delim);
 
+/** Join the source vector using 'delim' as the divider. */
+template<typename T>
```
Comment (abadams) [inline src/Util.h:190]: This is only going to work for strings now, so there's no point in templating it

Comment (steven-johnson) [reply src/Util.h:190]: could work for `char *` and `string_view` too

Comment (abadams) [reply src/Util.h:190]: Fair enough. It requires a .size() so not char *, but I guess it would work with string_view

**Pull request merged**